### PR TITLE
usb: add Chapter 9 (USB Device Framework) header

### PR DESCRIPTION
**Pull request by jfischer-no**
Add header file where all defines and structures
from Chapter 9 (USB Device Framework) should be included.

Replace all macros and types in the tree with the new ones from
usb/usb_ch9.h header.

Deprecate usb_common.h and usbstruct.h headers.
Deprecate DESCRIPTOR_TYPE_BOS macro.

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]

File: include/usb/usbstruct.h [hunk 1]
```diff
@@ -28,6 +28,12 @@
  *  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+/*
+ * This header and macros below are deprecated in 2.7 release.
+ * Please replace with macros from Chapter 9 header, include/usb/usb_ch9.h
+ */
+#warning "<usb/usbstruct.h> header is deprecated, use <usb/usb_ch9.h> instead"
```
Comment (mbolivar-nordic) [inline include/usb/usbstruct.h:35]: Shouldn't this be inside of the include guard?

**Comment by cfriedt** [comment]
@jfischer-no - could you fix your commit message? Is there anything else preventing this from being merged?

**Comment by mbolivar-nordic** [comment]
>  Is there anything else preventing this from being merged?

I think @jfischer-no should at least respond to the portability concerns I outlined.

File: include/usb/usb_ch9.h [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+#define USB_SFS_TEST_MODE		0x02
+
+/** Bits used for GetStatus response defined in spec. Figure 9-4 */
+#define USB_GET_STATUS_SELF_POWERED	BIT(0)
+#define USB_GET_STATUS_REMOTE_WAKEUP	BIT(1)
+
+static inline bool usb_reqtype_is_to_host(struct usb_setup_packet *setup)
+{
+	return setup->RequestType.direction == USB_REQTYPE_DIR_TO_HOST ?
+		true : false;
```
[the remark this answers is not in the document]
Comment (jfischer-no) [reply include/usb/usb_ch9.h:115]: I however find it easier to read, even if one need it inverted, `return foo == bar ? false : true;`

> Is there some MISRA thing where that requires a cast or something?

I do not know about MISRA. The type of 'foo == bar' is int, value 1 or 0, just as bool (_Bool) can only have these values.

File: include/usb/usb_ch9.h [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+	uint8_t recipient : 5;
+	uint8_t type : 2;
+	uint8_t direction : 1;
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+	uint8_t direction : 1;
+	uint8_t type : 2;
+	uint8_t recipient : 5;
+#endif
+} __packed;
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply include/usb/usb_ch9.h:35]: @mbolivar-nordic while you are entirely right in principle, multiple protocol stacks (including Bluetooth) use this style of bitfield declaration, and so does Linux and other close-to-metal open source projects. In practice all compilers respect this idiom, and so I would be in favor of keeping it here as well.

File: include/usb/usbstruct.h (hunk 1 above)
Comment (mbolivar-nordic) [reply include/usb/usbstruct.h:35]: > Shouldn't this be inside of the include guard?

Ping on this one

File: include/usb/usb_ch9.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply include/usb/usb_ch9.h:115]: I think it is verbose and harder to read, but OK, you're the maintainer

File: include/usb/usb_ch9.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbolivar-nordic) [reply include/usb/usb_ch9.h:35]: > In practice all compilers respect this idiom

If that is true, then okay.

> Given that using bitfields gives the compiler the opportunity to optimize, I would be in favor of keeping them.

Here I think a citation is needed that this actually results in better optimized code.

**Review by mbolivar-nordic — APPROVED** [review]

File: include/usb/usb_ch9.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply include/usb/usb_ch9.h:115]: I will change it.

File: include/usb/usbstruct.h (hunk 1 above)
Comment (jfischer-no) [reply include/usb/usbstruct.h:35]: I want to test something, I will remove or justify it at the next push.

File: include/usb/usb_ch9.h [hunk 4]
```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2020 PHYTEC Messtechnik GmbH
```
Comment (emob-nordic) [inline include/usb/usb_ch9.h:2]: Is it still valid?

File: include/usb/usb_ch9.h [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief USB Chapter 9 structures and definitions
+ *
+ * This file contains the USB Chapter 9 structures definitions
+ * and follows, with few exceptions, the USB Specification 2.0.
```
Comment (emob-nordic) [inline include/usb/usb_ch9.h:12]: You could mention those exceptions here.

**Review by eobalski — APPROVED** [review]

**Comment by cfriedt** [comment]
@jfischer-no - please mark any comments resolved if they are resolved, fix the merge conflict, and any compliance issues. I think everyone wants this to be in, so should be ready to merge when all of that is done.

File: include/usb/usb_ch9.h (hunk 4 above)
Comment (jfischer-no) [reply include/usb/usb_ch9.h:2]: Yes, this commit is from #27045

File: include/usb/usb_ch9.h (hunk 5 above)
Comment (jfischer-no) [reply include/usb/usb_ch9.h:12]: I will improve the documentation in subsequent PR's.

File: include/usb/usbstruct.h (hunk 1 above)
Comment (jfischer-no) [reply include/usb/usbstruct.h]: Moved inside include guard

**Pull request merged**